src/main/scala/is/hail/methods/Skat.scala

```diff
@@ -224,26 +227,33 @@ object Skat {
       (keyType, (keys: Annotation) => keys.asInstanceOf[Iterable[Annotation]].iterator)
     }
     
-    val nSamples = completeSamplesIndex.length
+    val nSamples = completeSampleIndex.length
+
+    val sc = vsm.sparkContext
+
+    val localGlobalAnnotationBc = sc.broadcast(vsm.globalAnnotation)
+    val sampleIdsBc = vsm.sampleIdsBc
+    val sampleAnnotationsBc = vsm.sampleAnnotationsBc
 
-    val sampleMask = new Array[Boolean](vds.nSamples)
-    completeSamplesIndex.foreach(i => sampleMask(i) = true)
-   
-    val sampleMaskBc = vds.sparkContext.broadcast(sampleMask)    
-    val completeSamplesBc = vds.sparkContext.broadcast(completeSamplesIndex)
+    val completeSampleIndexBc = sc.broadcast(completeSampleIndex)
 
-    (vds.rdd.flatMap { case (_, (va, gs)) =>
+    (vsm.rdd.flatMap { case (v, (va, gs)) =>
       (Option(keysQuerier(va)), typedWeightQuerier(va)) match {
         case (Some(key), Some(w)) =>
           if (w < 0)
             fatal(s"Variant weights must be non-negative, got $w")
-          val x: Vector[Double] =
-            if (useDosages) {
-              RegressionUtils.dosages(gs, completeSamplesBc.value)
-            } else {
-              RegressionUtils.hardCalls(gs, nSamples, sampleMaskBc.value)
-            }
-          keyIterator(key).map((_, (x, w)))
+
+          val n = completeSampleIndexBc.value.length
+          val x = new DenseVector[Double](n)
+
+          val missingSamples = new ArrayBuilder[Int]()
+
+          RegressionUtils.exprDosages(x,
```

Comment:
`exprDosages` seems like a bad name now.

Reply:
I agree.  In the next PR in this stack, I remove all the other dosage commands.  I'm going to rename it there as `inputVector`.